Image builder GitHub workflow integration design proposal. (#9970)
kyma-project/test-infra: cmd/image-builder/github-workflow-integration.md

Because OIDC token uses `JWT` format, it can be validated with a standard validation process against GitHub OIDC identity provider.

### Workflow Identification Claims

The OIDC token contains the following claims that can be used to identify the workflow that triggered the build pipeline. These include:

- `iss`: The issuer of the token. This is always `https://token.actions.githubusercontent.com`.
- `iat`: The time at which the token was issued.
- `exp`: The time at which the token expires.
- `jti`: A unique identifier for the token.
- `nbf`: The time before which the token must not be accepted for processing.
- `kid`: The key ID of the key used to sign the token.
- `alg`: The algorithm used to sign the token.
- `run_id`: The ID of the workflow run.
- `run_number`: The number of the workflow run.
- `actor`: The login of the user who initiated the workflow run.
- `event_name`: The name of the event that triggered the workflow run.
- `workflow`: The name of the workflow that triggered the workflow run.
- `workflow_ref`: The git ref associated with the workflow file.
- `repository`: The repository where the workflow run occurred.

> - `repository_owner`: The owner of the repository where the workflow run occurred.

Reviewer: Wouldn't the present form `occurs` be better in these descriptions?

> - `repository_owner`: The owner of the repository where the workflow run occurred.

Reviewer: If you can provide values for these claims, please use bold font for them (like you do for [parameters and variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font)).

Author: Values for claims are set by GitHub OIDC identity provider.

### Source Code Cloning Claims

The OIDC token also contains claims that can be used to clone the appropriate version of the source code:

- `repository`: The repository where the workflow run occurred.
- `repository_owner`: The owner of the repository where the workflow run occurred.

> - `event_name`: The name of the event that triggered the workflow run.

Reviewer: Please consider using the present forms of the verbs.

> - `event_name`: The name of the event that triggered the workflow run.

Reviewer: If you can provide values for these claims, please use bold font for them (like you do for [parameters and variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font)).

Author: Values for claims are set by GitHub OIDC identity provider.

- `ref`: The git ref associated with the workflow run. This can be used to checkout the correct branch, tag, or commit.
- `ref_type`: The type of git ref associated with the workflow run. This can be used to determine whether the ref is a branch, tag, or
  commit.
- `base_ref`: The base git ref associated with the workflow run. This can be used to determine the base branch for a pull request.

> - `head_ref`: The head git ref associated with the workflow run. This can be used to determine the head branch for a pull request.

Reviewer: If you can provide values for these claims, please use bold font for them (like you do for [parameters and variables](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#code-font-and-bold-font)).

Reviewer: What I meant is, that claims are similar to parameters or variables In the sense that they can have some values (no matter where the values come from). Hence, they should be formatted in the same way, i.e. in bold font.

These claims ensure that the `oci-image-builder` pipeline builds the exact version of the code that was provided in the PR or merged to the

> branch, adhering to slc-29 compliance.

Reviewer: ```suggestion
branch, adhering to SLC-29 compliance.
```

## Parameters for the `oci-image-builder` Pipeline

The `oci-image-builder` pipeline requires certain data to be provided in parameters to execute a build process.

> Certain parameters need to be defined by the user additionally to the data taken from OIDC token.

Reviewer: ```suggestion
Certain parameters need to be defined by the user in addition to the data taken from the OIDC token.
```

> These parameters include:

Reviewer: ```suggestion
These parameters include user-defined parameters, parameters from the OIDC token, and computed parameters.
```

### User Defined Parameters

- `Context`: The context of the build.
- `Dockerfile`: The Dockerfile to be used for the build.
- `Name`: The name of the image.
- `BuildArgs`: The build arguments to be passed to the build.
- `Tags`: The tags to be applied to the image.

> - `ExportTags`: Whether to export the tags.

Reviewer 2: What does it mean to export the tag?

Author: From image-builder.

```
        flagSet.BoolVar(&o.exportTags, "export-tags", false, "Export parsed tags as build-args into dockerfile. Each tag will have format TAG_x, where x is the tag name passed along with the tag")
```

> - `Context`: The context of the build.
> - `Dockerfile`: The Dockerfile to be used for the build.
> - `Name`: The name of the image.
> - `BuildArgs`: The build arguments to be passed to the build.
> - `Tags`: The tags to be applied to the image.
> - `ExportTags`: Whether to export the tags.

Reviewer: ```suggestion
- **Context**: The context of the build.
- **Dockerfile**: The Dockerfile to be used for the build.
- **Name**: The name of the image.
- **BuildArgs**: The build arguments to be passed to the build.
- **Tags**: The tags to be applied to the image.
- **ExportTags**: Whether to export the tags.
```

### Parameters from OIDC Token

- `RepoName`: The name of the repository.

> - `RepoOwner`: The owner of the repository. Possible values include "kyma-project" and "kyma-incubator".

Reviewer 2: kyma-incubator will be owned by compass team and I am not sure if they should be included into new image builder as they should migrate out of Prow until end of June. I think this is question for @TorstenD-SAP .

Author: When `kyma-incubator` org will not be allowed to build images with our solution, we will simply remove respective value from allowed org list.

> - `RepoName`: The name of the repository.
> - `RepoOwner`: The owner of the repository. Possible values include "kyma-project" and "kyma-incubator".

Reviewer: ```suggestion
- **RepoName**: The name of the repository.
- **RepoOwner**: The owner of the repository. Possible values include `kyma-project` and `kyma-incubator`.
```

> - `JobType`: The type of job. Possible values include "presubmit" and "postsubmit".
> - `PullBaseSHA`: The base SHA of the pull request.
> - `PullPullSHA`: The SHA of the pull request.

Reviewer: ```suggestion
- **JobType**: The type of job. Possible values include `presubmit` and `postsubmit`.
- **PullBaseSHA**: The base SHA of the pull request.
- **PullPullSHA**: The SHA of the pull request.
```

### Computed Parameters

> The data for this parameter is not available in OIDC token and should not be provided by user.

Reviewer: ```suggestion
The data for this parameter is not available in the OIDC token and should not be provided by the user.
```

> We can extract it from `ref` claim of OIDC token.

Reviewer: ```suggestion
We can extract it from the `ref` claim of the OIDC token.
```

> - `PullNumber`: The number of the pull request.

Reviewer: ```suggestion
- **PullNumber**: The number of the pull request.
```

## ProwJob Validation in Image Builder

In the scenario where the `oci-image-builder` pipeline is triggered from a ProwJob instead of a GitHub workflow, the validation process

> differs slightly due to the absence of an OIDC token. Triggering the `oci-image-builder` pipeline from a ProwJob is considered a depraecated

Reviewer: ```suggestion
differs slightly due to the absence of an OIDC token. Triggering the `oci-image-builder` pipeline from a ProwJob is considered a deprecated
```

scenario and must be supported only for backward compatibility and migration purposes.

### Process Flow

> 1. **Trigger ProwJob**: User or automation triggers a ProwJob.

Reviewer: ```suggestion
1. **Trigger ProwJob**: The user or automation triggers a ProwJob.
```

> 2. **Trigger oci-image-builder pipeline**: The ProwJob running Image builder client calls ADO API to trigger `oci-image-builder` pipeline.

Reviewer: ```suggestion
2. **Trigger oci-image-builder pipeline**: The ProwJob running Image builder client calls ADO API to trigger the `oci-image-builder` pipeline.
```

>    The parameters required by the `oci-image-builder` pipeline are passed as parameters to the pipeline. These parameters values are
>    collected from data defined by the user and Prow environment variables. Together with the parameters, parameters hash signed by Prow is
>    also passed to the pipeline as a parameter.

Reviewer: ```suggestion
   The parameters required by the `oci-image-builder` pipeline are passed as parameters to the pipeline. These parameters' values are
   collected from data defined by the user and Prow environment variables. Together with these parameters, the parameters' hash signed by Prow is
   also passed to the pipeline as a parameter.
```

> 3. **Validating the Parameters**: The `oci-image-builder` pipeline, running in Azure DevOps, detects the call originates from Prow and

Reviewer: ```suggestion
3. **Validating the Parameters**: The `oci-image-builder` pipeline, running in Azure DevOps, detects the call originating from Prow and
```

   validates the parameters against a hash signed by Prow. This step ensures that the parameters are valid and have not been tampered with.
   The absence of an OIDC token is used to detect that the call is coming from a ProwJob.

### Parameters Validation

> The parameters passed to the `oci-image-builder` pipeline are validated against a hash signed by Prow. This hash is computed based on the
> names and values of the parameters. The hash value is signed with secret know to the `Prow` and `oci-image-buidler` pipeline only.

Reviewer: ```suggestion
The parameters passed to the `oci-image-builder` pipeline are validated against a hash signed by Prow. This hash is computed based on the
names and values of the parameters. The hash value is signed with a secret known to the `Prow` and the `oci-image-buidler` pipeline only.
```

The `oci-image-builder` pipeline verifies this hash to ensure that the parameters have not been tampered with and are indeed coming from the
ProwJob.

This validation process ensures the secure and authorized passing of information about the ProwJob and the image to build. It provides a
reliable infrastructure for the building of OCI images when the pipeline is triggered from a ProwJob.

## Block Diagram

![image-builder-block-diagram](documentation_assets/image-builder-block-diagram.png)

## Activity Diagram

![image-builder-activity-diagram](documentation_assets/image-builder-activity-diagram.png)

## Conclusion

The Image Builder solution, with its seamless integration with GitHub workflows and Azure DevOps pipeline, offers developers a robust and
secure method to incorporate the building of OCI images into their workflows. By leveraging a signed JWT format in which an OIDC token from
GitHub's OIDC identity provider is passed, it ensures the secure and authorized passing of information about the workflow and the image to

> build. The entire build process adheres to slc-29 compliance, providing a reliable infrastructure for the building of OCI images.

Reviewer: ```suggestion
build. The entire build process adheres to SLC-29 compliance, providing a reliable infrastructure for the building of OCI images.
```

8. **Signing the OCI Image**: If the build was triggered by a push event in GitHub, the `oci-image-builder` pipeline uses the `signify`
   service to sign the OCI image.
   This step ensures the integrity and authenticity of the OCI image.

## GitHub OIDC Identity Token Claims

The OIDC token issued by GitHub's OIDC identity provider contains several claims that are crucial for the `oci-image-builder` pipeline.
These claims are used to identify the workflow triggering the build pipeline and to clone the appropriate version of the source code. This
is essential for SLC-29 compliance, as it ensures that the exact version of the code that was tested in the PR or for which the push was
merged is built.

The validity and integrity of the OIDC token must be validated in the `oci-image-builder` pipeline and fail the pipeline execution if validation fails.
Because the OIDC token uses the `JWT` format, it can be validated with a standard validation process against GitHub's OIDC identity provider.

### Workflow Identification Claims

The OIDC token contains the following claims that can be used to identify the workflow that triggered the build pipeline. These include:

> Sure, here's how you can format the selected text as an unordered list in Markdown, with the words before the colon surrounded by backticks:

Reviewer: I guess this isn't part of the doc.

<!-- markdown-link-check-disable -->

- `iss`: The issuer of the token. This is always https://token.actions.githubusercontent.com. <!-- markdown-link-check-enable-->
- `iat`: The time when the token is issued.
- `exp`: The time when the token expires.
- `jti`: A unique identifier for the token.
- `nbf`: The time before which the token must not be accepted for processing.
- `kid`: The key ID of the key used to sign the token.
- `alg`: The algorithm used to sign the token.
- `run_id`: The ID of the workflow run.
- `run_number`: The number of the workflow run.
- `actor`: The login of the user who initiates the workflow run.
- `event_name`: The name of the event that triggers the workflow run.
- `workflow`: The name of the workflow that triggers the workflow run.
- `workflow_ref`: The git ref associated with the workflow file.
- `repository`: The repository where the workflow run occurs.
- `repository_owner`: The owner of the repository where the workflow run occurs.

### Source Code Cloning Claims

The OIDC token also contains claims that can be used to clone the appropriate version of the source code:

> Sure, here's the selected text rewritten with present tense verbs, while keeping the formatting:

Reviewer: Same here - the comment doesn't belong here.
